### PR TITLE
Implement attribute placement validators and name-uniqueness checks

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds plan-level validators for attribute placement and name uniqueness in the DDlog parser.
- Introduces an attribute span scanner, integrates it into the parsing pipeline, and adds a post-parse validator to enforce uniqueness across top-level items.
- Updates tests and documentation to cover new behaviour.

## Changes
- New attribute span scanner and integration
  - Added src/parser/span_scanners/attributes.rs to collect attribute spans and validate their placement.
  - Wired the attribute scanner into the parsing flow:
    - Updated src/parser/span_scanner.rs to collect attribute spans, extend non-rule span capacity, and attach attributes to ParsedSpans.
    - Extended src/parser/span_scanners/mod.rs to expose collect_attribute_spans.
    - Updated src/parser/cst_builder/spans.rs and src/parser/cst_builder/tree.rs to include and map N_ATTRIBUTE in the CST tree.
  - Updated src/parser/span_scanner.rs to merge attribute spans into the parsing pipeline and surface any attribute-related errors.
- Name uniqueness validator (post-parse)
  - Added src/parser/validators/name_uniqueness.rs implementing:
    - duplicate type, relation, index, transformer, and import name checks
    - function-name-and-arity (arity-based overloading allowed; duplicates are errors)
    - text_range_to_span-based error spans
  - Added src/parser/validators/mod.rs to host validators and wired into the parser.
  - Wired into the main parse() path in src/parser/mod.rs so errors from validation appear in final Parsed.
- AST/ParsedSpans updates
  - Modified src/parser/cst_builder/spans.rs to add attributes: Vec<Span> to ParsedSpans and its Builder, including accessors and span-list registration.
  - Updated src/parser/cst_builder/tree.rs SPAN_CURSOR_KINDS to include (ParsedSpans::attributes, SyntaxKind::N_ATTRIBUTE).
- Tests
  - Unit tests for attribute spans: src/parser/span_scanners/tests.rs (collect_attribute_spans tests).
  - Parser-level tests for attributes: src/parser/tests/attributes.rs and mod registration in src/parser/tests/mod.rs.
  - Integration/behavioural tests: tests/attribute_placement.rs and tests/name_uniqueness.rs.
  - Name-uniqueness unit tests added within the new name_uniqueness.rs module tests.
- Documentation and roadmap
  - Added docs/execplans/phase-2-attribute-placement-validators.md detailing the validators, design, and plan.
  - Updated docs/roadmap.md to mark the validators item as completed (x).

## Implementation details
- Attribute scanner design
  - Detects attributes beginning with # and [; records attribute spans; enforces that attributes target only permitted hosts (typedef, type, function, extern, input/output/relational/stream/multiset variants).
  - Emits a diagnostic when an attribute is placed on a non-permitted item, with a range covering the offending attribute(s).
  - Uses existing chumsky Simple<SyntaxKind> error format for consistency with other parsers.
- Name uniqueness validator
  - Traverses the AST root and checks per-category duplicates: types, relations, indexes, transformers, imports.
  - Functions checked by (name, arity) pairs; arity-based overloading is allowed per spec.
  - Uses text_range_to_span to produce accurate error spans.
  - Malformed items (missing names) are skipped to avoid false positives.
- Testing approach
  - Unit tests focus on the attribute scanner’s span collection and error generation.
  - Parser-level tests exercise the parse() pipeline for various attribute placements.
  - Integration tests validate real-world usage and cross-item interactions; include positive and negative cases.
  - Name-uniqueness tests cover duplicates across categories and function overloading rules.

## Testing plan
- Run the repository’s gates locally:
  - set -o pipefail && make check-fmt 2>&1 | tee /tmp/ddlint-check-fmt.log
  - set -o pipefail && make lint 2>&1 | tee /tmp/ddlint-lint.log
  - set -o pipefail && make test 2>&1 | tee /tmp/ddlint-test.log
- Validate acceptance criteria from the ExecPlan:
  - Attribute placement and duplicate-name diagnostics are produced as specified.
  - Attribute spans are reflected in CST wrappers as N_ATTRIBUTE nodes when parsing.

## Why this approach
- Aligns attribute handling with the existing parsing architecture (span scanners -> CST -> AST wrappers) for clean, uniform processing.
- Keeps the changes localized to existing modules without introducing new crates or breaking public APIs.
- Provides a robust, whole-program validation via a post-parse pass for name uniqueness, enabling cross-item consistency checks.

## Documentation / Roadmap
- Phase-2 plan doc added: docs/execplans/phase-2-attribute-placement-validators.md
- Roadmap updated: docs/roadmap.md now marks the validators item as complete
  - [x] Add validators for attribute placement and name uniqueness to mirror the specification's guarantees.

## Notes for reviewers
- The tests rely on Debug formatting for error inspection (Simple<SyntaxKind>), consistent with existing tests.
- Some tests demonstrate known parser limitations (e.g., consecutive relation declarations merging spans); tests intentionally interleave items to work around this.
- No external dependencies were added; all changes stay within existing crate boundaries.

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---

ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/b21302c9-99bd-4107-aee3-c172fec2e107

## Summary by Sourcery

Add parser support for collecting attribute spans and validating their placement, and introduce post-parse name-uniqueness validation across top-level declarations.

Enhancements:
- Extend ParsedSpans and CST construction to track attribute spans as N_ATTRIBUTE nodes in the syntax tree.
- Integrate a new attribute span scanner into the parsing pipeline so attribute-related diagnostics are produced alongside other parse errors.
- Add a name-uniqueness validator over the AST to enforce unique names for types, relations, indexes, transformers, imports, and function (name, arity) pairs.

Documentation:
- Document the design and behaviour of attribute-placement and name-uniqueness validators and mark the corresponding roadmap item as complete.

Tests:
- Add unit, parser-level, and behavioural tests covering attribute span collection, attribute placement rules, and name-uniqueness validation.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[src/parser/validators/name_uniqueness.rs](https://github.com/leynos/ddlint/pull/209/files/6fb332c357efe645f7ee0da6d0bc28ee9d3e6553#diff-d09a40ca07ac9e064c1e3e90e41d27a803d718eb5f4ff75598798b3028ad4e7c)

Comment on lines +162 to +172

~~~rust
    fn duplicate_typedef_names() {
        let src = "typedef A = u32\ntypedef A = string";
        let parsed = parse(src);
        let errors = super::validate_name_uniqueness(parsed.root());
        assert_eq!(errors.len(), 1);
        let e = errors.first().expect("expected one error");
        assert!(
            render(e).contains("duplicate type name 'A'"),
            "unexpected error: {e:?}",
        );
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/69278/delta?repo-id=1160463&review-id=209&biomarker=Code+Duplication&filename=src%2Fparser%2Fvalidators%2Fname_uniqueness.rs&method=)
The module contains 5 functions with similar structure: tests.duplicate_function_name_and_arity,tests.duplicate_import_paths,tests.duplicate_index_names,tests.duplicate_transformer_names and 1 more functions

**Comment by leynos** [comment]
@coderabbitai have the following now been resolved?

~~~
Please address the comments from this code review:

## Overall Comments
- The attribute placement check currently treats `K_EXTERN` itself as a permitted target without peeking ahead, so constructs like `#[cold]
extern index I ...` (if ever allowed by the grammar) would be incorrectly accepted; consider tightening `is_permitted_attribute_target` to require `extern` to be followed specifically by `type`/`function` before treating it as valid.
- In `handle_hash`, all collected attribute spans (including those on forbidden targets) are still added to `ParsedSpans` and thus appear as `N_ATTRIBUTE` nodes; if the intention is that invalid attributes should not participate in the CST, consider omitting those spans when an error is emitted or otherwise marking them so downstream passes can distinguish them.

## Individual Comments

### Comment 1
<location> `src/parser/span_scanners/attributes.rs:88-97` </location>
<code_context>
+fn handle_hash(st: &mut State<'_>, _span: Span) {
</code_context>

<issue_to_address>
**nitpick:** Avoid building a dummy 0-length span for the placement error when `attr_spans` cannot be empty.

Since `attr_spans` must be non-empty here (we early-return on `consume_attribute` failure), the `unwrap_or_default()` calls for `first_start` and `last_end` are effectively dead paths and would silently yield a `0..0` span only if that invariant were broken. It would be clearer to either unwrap directly or assert non-emptiness so that any invariant violation is surfaced explicitly.
</issue_to_address>

### Comment 2
<location> `src/parser/span_scanners/tests.rs:392-401` </location>
<code_context>
+    clippy::expect_used,
+    reason = "tests unwrap errors to provide crisp failure messages"
+)]
+#[test]
+fn collect_attribute_spans_rejected_on_index() {
+    let src = "#[cold]\nindex I on R(x)\n";
+    let tokens = tokenize(src);
+    let (spans, errors) = collect_attribute_spans(&tokens, src);
+    assert_eq!(spans.len(), 1);
+    assert_eq!(errors.len(), 1);
+    let first_error = errors.first().expect("expected at least one error");
+    assert!(
+        format!("{first_error:?}").contains("attribute"),
+        "unexpected error: {first_error:?}",
+    );
+}
+
+#[test]
+fn collect_attribute_spans_rejected_on_apply() {
+    let src = "#[cold]\napply T(R) -> (S)\n";
+    let tokens = tokenize(src);
+    let (spans, errors) = collect_attribute_spans(&tokens, src);
+    assert_eq!(spans.len(), 1);
+    assert_eq!(errors.len(), 1);
+}
+
+#[test]
+fn collect_attribute_spans_rejected_on_import() {
+    let src = "#[cold]\nimport foo\n";
+    let tokens = tokenize(src);
</code_context>

<issue_to_address>
**suggestion (testing):** Tighten attribute scanner negative tests and add a case for unclosed attribute brackets

For the apply/import tests, consider also asserting that the single diagnostic is specifically about attribute placement (e.g., message contains "attribute"), mirroring the index test, so we know we're not just seeing some unrelated parse error. Also, there’s no coverage for the `unclosed attribute bracket` branch in `consume_attribute`; a test like `"#[cold"` that expects one error with a message containing "unclosed attribute bracket" would exercise that path and protect the bracket-matching logic from regressions.

Suggested implementation:

```rust
#[test]
fn collect_attribute_spans_rejected_on_apply() {
    let src = "#[cold]\napply T(R) -> (S)\n";
    let tokens = tokenize(src);
    let (spans, errors) = collect_attribute_spans(&tokens, src);
    assert_eq!(spans.len(), 1);
    assert_eq!(errors.len(), 1);

    let [first_error] = &errors[..] else {
        panic!("expected exactly one error, got {errors:?}");
    };
    assert!(
        format!("{first_error:?}").contains("attribute"),
        "unexpected error: {first_error:?}",
    );
}

```

```rust
#[test]
fn collect_attribute_spans_rejected_on_import() {
    let src = "#[cold]\nimport foo\n";
    let tokens = tokenize(src);
    let (spans, errors) = collect_attribute_spans(&tokens, src);
    assert_eq!(spans.len(), 1);
    assert_eq!(errors.len(), 1);

    let [first_error] = &errors[..] else {
        panic!("expected exactly one error, got {errors:?}");
    };
    assert!(
        format!("{first_error:?}").contains("attribute"),
        "unexpected error: {first_error:?}",
    );
}

#[test]
fn collect_attribute_spans_unclosed_attribute_bracket() {
    let src = "#[cold";
    let tokens = tokenize(src);
    let (spans, errors) = collect_attribute_spans(&tokens, src);

    // Attribute should not be collected as a valid span when the bracket is unclosed.
    assert_eq!(spans.len(), 0);
    assert_eq!(errors.len(), 1);

    let [first_error] = &errors[..] else {
        panic!("expected exactly one error, got {errors:?}");
    };
    assert!(
        format!("{first_error:?}").contains("unclosed attribute bracket"),
        "unexpected error: {first_error:?}",
    );
}

```
</issue_to_address>

### Comment 3
<location> `src/parser/tests/attributes.rs:11-19` </location>
<code_context>
+use crate::test_util::assert_no_parse_errors;
+use rstest::rstest;
+
+#[rstest]
+#[case("#[cold]\ntypedef T = u32")]
+#[case("#[inline]\nfunction f() {}")]
+#[case("#[hot]\ninput relation R(x: u32)")]
+#[case("#[cold]\noutput relation R(x: u32)")]
+#[case("#[cold]\nrelation R(x: u32)")]
+#[case("#[cold]\nextern function f()")]
+#[case("#[cold]\nextern type Handle")]
+fn attribute_on_permitted_item_no_error(#[case] src: &str) {
+    let parsed = parse(src);
</code_context>

<issue_to_address>
**suggestion (testing):** Add tests for attributes on `stream` and `multiset` relations to match the permitted targets

The implementation permits attributes on `stream` and `multiset` relations (`is_permitted_attribute_target` matches `K_STREAM` and `K_MULTISET`), but the tests here only cover `input`, `output`, and bare `relation`. Please add cases like `"#[hot]\nstream relation R(x: u32)"` and `"#[hot]\nmultiset relation R(x: u32)"` to `attribute_on_permitted_item_no_error` so the parser tests cover all permitted relation forms.

```suggestion
#[rstest]
#[case("#[cold]\ntypedef T = u32")]
#[case("#[inline]\nfunction f() {}")]
#[case("#[hot]\ninput relation R(x: u32)")]
#[case("#[hot]\nstream relation R(x: u32)")]
#[case("#[hot]\nmultiset relation R(x: u32)")]
#[case("#[cold]\noutput relation R(x: u32)")]
#[case("#[cold]\nrelation R(x: u32)")]
#[case("#[cold]\nextern function f()")]
#[case("#[cold]\nextern type Handle")]
fn attribute_on_permitted_item_no_error(#[case] src: &str) {
```
</issue_to_address>

### Comment 4
<location> `tests/attribute_placement.rs:11-20` </location>
<code_context>
+use crate::test_util::assert_no_parse_errors;
+use rstest::rstest;
+
+#[rstest]
+#[case("#[cold]\ntypedef T = u32")]
+#[case("#[inline]\nfunction f() {}")]
+#[case("#[hot]\ninput relation R(x: u32)")]
+#[case("#[cold]\noutput relation R(x: u32)")]
+#[case("#[cold]\nrelation R(x: u32)")]
+#[case("#[cold]\nextern function f()")]
+#[case("#[cold]\nextern type Handle")]
+fn attribute_on_permitted_item_no_error(#[case] src: &str) {
+    let parsed = parse(src);
</code_context>

<issue_to_address>
**suggestion (testing):** Mirror stream/multiset relation attribute cases in behavioural tests

Please also add `stream` and `multiset` relation examples (e.g. `#[hot]
stream relation R(x: u32)` and `#[hot]
multiset relation R(x: u32)`) to `valid_attribute_placement` so behavioural tests stay in sync with the parser tests and the implementation’s allowed targets, and will catch any future regressions.

```suggestion
#[rstest]
#[case("#[cold]\ntypedef T = u32")]
#[case("#[inline]\nfunction f() {}")]
#[case("#[hot]\ninput relation R(x: u32)")]
#[case("#[hot]\noutput relation R(x: u32)")]
#[case("#[hot]\nrelation R(x: u32)")]
#[case("#[hot]\nstream relation R(x: u32)")]
#[case("#[hot]\nmultiset relation R(x: u32)")]
#[case("#[cold]\nextern function f()")]
#[case("#[cold]\nextern type Handle")]
#[case("#[a]\n#[b]\ntypedef T = u32")]
fn valid_attribute_placement(#[case] src: &str) {
```
</issue_to_address>

### Comment 5
<location> `tests/name_uniqueness.rs:10-19` </location>
<code_context>
+    clippy::expect_used,
+    reason = "tests unwrap spans to provide crisp failure messages"
+)]
+#[test]
+fn collect_attribute_spans_valid_on_typedef() {
+    let src = "#[cold]\ntypedef T = u32\n";
</code_context>

<issue_to_address>
**suggestion (testing):** Consider adding an integration test for multiple duplicate definitions in one program

Current tests check each duplicate type (typedef, relation, index, transformer, import, function) in isolation and confirm that function overloading by arity is allowed. To better cover the end-to-end `parse()` path, add a program that includes several different duplicate-name violations at once and assert that multiple diagnostics are emitted, demonstrating that `validate_name_uniqueness` accumulates errors instead of short-circuiting.
</issue_to_address>

### Comment 6
<location> `docs/execplans/phase-2-attribute-placement-validators.md:168-177` </location>
<code_context>
+Summary of delivered artefacts:
</code_context>

<issue_to_address>
**suggestion (typo):** Spelling of "artefacts"/"artifacts" is inconsistent; consider standardising (e.g. to en-GB "artefacts").

Earlier you use en-GB "artefacts" ("Summary of delivered artefacts"), but later the heading is "Artifacts and notes". Since the Constraints specify en-GB-oxendict, consider changing that heading to "Artefacts and notes" (or otherwise choose one spelling and use it consistently).

Suggested implementation:

```
## Artefacts and notes

```

Please also:
1. Grep this file (and nearby execution plan docs) for the string `Artifacts` and change any remaining instances to `Artefacts` for consistency with en-GB-oxendict.
2. Ensure any cross-references or table-of-contents entries that mention “Artifacts and notes” are updated to “Artefacts and notes” so links and anchors remain correct.
</issue_to_address>

### Comment 7
<location> `docs/execplans/phase-2-attribute-placement-validators.md:3` </location>
<code_context>
+# Add validators for attribute placement and name uniqueness
+
+This ExecPlan is a living document. The sections `Constraints`, `Tolerances`,
+`Risks`, `Progress`, `Surprises & Discoveries`, `Decision Log`, and
+`Outcomes & Retrospective` must be kept up to date as work proceeds.
</code_context>

<issue_to_address>
**issue (review_instructions):** "ExecPlan" is used as an acronym without being defined on first use.

Please expand this to define the term on first use, for example "Execution Plan (ExecPlan) is a living document…" so that the acronym is clear to readers.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 8
<location> `docs/execplans/phase-2-attribute-placement-validators.md:13` </location>
<code_context>
+
+## Purpose / big picture
+
+After this change a user running `ddlint` on a DDlog program will see
+span-accurate diagnostics for two categories of mistake that the parser
+currently ignores:
</code_context>

<issue_to_address>
**issue (review_instructions):** "DDlog" appears here without its full form being introduced first.

Consider expanding this to something like "Differential Datalog (DDlog)" on first use so that the acronym is defined for readers unfamiliar with it.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 9
<location> `docs/execplans/phase-2-attribute-placement-validators.md:41` </location>
<code_context>
+
+- Follow `docs/differential-datalog-parser-syntax-spec-updated.md` sections
+  5.1, 8, 9, and 12 as the normative specification.
+- Preserve existing parser architecture: tokenizer -> span scanners ->
+  `ParsedSpans` builder -> CST green tree -> AST wrappers -> `Parsed`.
+- Do not add new crate dependencies.
</code_context>

<issue_to_address>
**issue (review_instructions):** CST and AST are introduced later in this sentence without being defined on first use.

In this bullet (or immediately following it), please spell out the acronyms on first use, for example "Concrete Syntax Tree (CST)" and "Abstract Syntax Tree (AST)", so that the architecture description is unambiguous.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 10
<location> `docs/execplans/phase-2-attribute-placement-validators.md:58` </location>
<code_context>
+
+- Scope: if implementation requires changes to more than 15 files or 600 net
+  new lines, stop and escalate.
+- Interface: if any existing public API signature (on `Parsed`, `Root`, or any
+  AST wrapper) must change, stop and escalate.
+- Dependencies: if a new external crate dependency is required, stop and
</code_context>

<issue_to_address>
**issue (review_instructions):** "API" is used here without expanding the acronym on first use.

Please expand "API" the first time it appears, e.g. "public Application Programming Interface (API) signature", to comply with the acronym definition guideline.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: src/parser/span_scanners/tests.rs [hunk 1]
```diff
@@ -384,3 +384,102 @@ fn parse_tokens_skips_non_rule_constructs_when_scanning_rules() {
...
+}
+
+#[test]
+fn collect_attribute_spans_rejected_on_import() {
+    let src = "#[cold]\nimport foo\n";
+    let tokens = tokenize(src);
+    let (spans, errors) = collect_attribute_spans(&tokens, src);
+    assert_eq!(spans.len(), 1);
+    assert_eq!(errors.len(), 1);
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/parser/span_scanners/tests.rs]: @coderabbitai has this now been resolved?

Use sourcegraph analysis to determine your answer.

If the comment has been resolved, please mark it as resolved using the API. Otherwise, please provide an AI agent prompt for the work remaining.

File: src/parser/span_scanners/tests.rs [hunk 2]
```diff
@@ -384,3 +384,102 @@ fn parse_tokens_skips_non_rule_constructs_when_scanning_rules() {
...
+
+#[test]
+fn collect_attribute_spans_valid_on_extern_function() {
+    let src = "#[cold]\nextern function f()\n";
+    let tokens = tokenize(src);
+    let (spans, errors) = collect_attribute_spans(&tokens, src);
+    assert_eq!(spans.len(), 1);
+    assert!(errors.is_empty(), "unexpected errors: {errors:?}");
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/parser/span_scanners/tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
